tracer/src/Datadog.Trace/OTelMetrics/TagSet.cs

```diff
@@ -19,22 +20,63 @@ namespace Datadog.Trace.OTelMetrics;
 /// </summary>
 internal readonly struct TagSet : IEquatable<TagSet>
 {
+    private static readonly IComparer<KeyValuePair<string, object?>> KeyComparer =
+        Comparer<KeyValuePair<string, object?>>.Create(static (a, b) => string.CompareOrdinal(a.Key, b.Key));
+
     private readonly string _key;
 
     private TagSet(string key) => _key = key;
 
     public static TagSet FromSpan(ReadOnlySpan<KeyValuePair<string, object?>> tags)
     {
-        if (tags.Length == 0)
+        var len = tags.Length;
+        if (len == 0)
         {
             return new TagSet(string.Empty);
         }
 
-        var sorted = tags.ToArray();
-        Array.Sort(sorted, (a, b) => string.CompareOrdinal(a.Key, b.Key));
+        if (len == 1)
+        {
+            ref readonly var kv = ref tags[0];
+            var sb = StringBuilderCache.Acquire();
+            sb.Append(kv.Key).Append('=');
+            if (kv.Value is not null)
+            {
+                sb.Append(kv.Value);
+            }
+
+            return new TagSet(StringBuilderCache.GetStringAndRelease(sb));
+        }
+
+        var sorted = ArrayPool<KeyValuePair<string, object?>>.Shared.Rent(len);
+        try
+        {
+            tags.CopyTo(sorted);
+            Array.Sort(sorted, 0, len, KeyComparer);
 
-        var key = string.Join(";", sorted.Select(kv => $"{kv.Key}={kv.Value}"));
-        return new TagSet(key);
+            var sb = StringBuilderCache.Acquire();
+            for (int i = 0; i < len; i++)
+            {
+                if (i > 0)
+                {
+                    sb.Append(';');
+                }
+
+                ref readonly var kv = ref sorted[i];
```

Comment:
Hm maybe I'm missing something as there is the `ref readonly` and `ref` here too as in the other location.

Why exactly do we need to do this? Probably worthwhile to note down as a comment as it isn't immediately clear to me.